File: sdk-python/impreza/resources/invoices.py
"""Invoices resource — accessed via ``Client.invoices``.

Read access (``list`` / ``get``) plus pay-from-balance (``pay``).

``pay`` spends real account credit and is not reversible from here, so
it takes no ``confirm`` flag by design — the decision belongs to the
caller, and every CLI/UI on top of this prompts before calling it.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..models.invoice import Invoice, InvoiceDetail, InvoicePayment

if TYPE_CHECKING:  # pragma: no cover
    from .._http import HttpClient
    from .._http_async import AsyncHttpClient


def _extract_invoices(payload: dict[str, object]) -> list[Invoice]:
    data_raw = payload.get("data")
    data = data_raw if isinstance(data_raw, dict) else {}
    items_raw = data.get("invoices")
    items = items_raw if isinstance(items_raw, list) else []
    return [Invoice.model_validate(item) for item in items]


def _extract_invoice_detail(payload: dict[str, object]) -> InvoiceDetail:
    data_raw = payload.get("data")
    data = data_raw if isinstance(data_raw, dict) else {}
    return InvoiceDetail.model_validate(data)


def _extract_payment(payload: dict[str, object], invoice_id: int) -> InvoicePayment:
    data_raw = payload.get("data")
    data = data_raw if isinstance(data_raw, dict) else {}
    # The endpoint echoes invoice_id, but fall back to the one we asked
    # about so the result is never ambiguous about which invoice moved.
    data.setdefault("invoice_id", invoice_id)
    return InvoicePayment.model_validate(data)


def _list_params(status: str | None) -> dict[str, object] | None:
    return {"status": status} if status else None


class InvoicesResource:
    """Sync access to the authenticated client's invoices."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self, *, status: str | None = None) -> list[Invoice]:
        """List invoices, optionally filtered by status.

        ``status`` accepts these values: ``Paid``, ``Unpaid``, ``Cancelled``,
        ``Refunded``.
        """
        payload = self._http.get("/invoices", params=_list_params(status))
        return _extract_invoices(payload)

    def get(self, invoice_id: int) -> InvoiceDetail:
        """Return one invoice with its line items and transaction history."""
        payload = self._http.get(f"/invoices/{invoice_id}")
        return _extract_invoice_detail(payload)

    def pay(self, invoice_id: int) -> InvoicePayment:
        """Pay an unpaid invoice from the account credit balance.

        This spends real money and cannot be undone through the API.

        Raises:
            Conflict: 409 — the invoice is already paid
                (``code == "ALREADY_PAID"``) or the balance does not cover
                it (``code == "INSUFFICIENT_BALANCE"``). Top up first with
                ``client.account.topup(...)``.
            ResourceNotFound: 404 — no such invoice on this account.
        """
        payload = self._http.post(f"/invoices/{invoice_id}/pay")
        return _extract_payment(payload, invoice_id)


class AsyncInvoicesResource:
    """Async access to the authenticated client's invoices."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def list(self, *, status: str | None = None) -> list[Invoice]:
        payload = await self._http.get("/invoices", params=_list_params(status))
        return _extract_invoices(payload)

    async def get(self, invoice_id: int) -> InvoiceDetail:
        payload = await self._http.get(f"/invoices/{invoice_id}")
        return _extract_invoice_detail(payload)

    async def pay(self, invoice_id: int) -> InvoicePayment:
        """Pay an unpaid invoice from the account credit balance.

        Spends real money; see :meth:`InvoicesResource.pay`.
        """
        payload = await self._http.post(f"/invoices/{invoice_id}/pay")
        return _extract_payment(payload, invoice_id)

File: sdk-go/client/invoices.go
package client

import (
	"context"
	"net/url"
)

// Invoice is one row of GET /v1/invoices. Field shape matches the
// live response: invoice_num is the customer-facing reference (not
// to be confused with id which is the WHMCS-internal primary key).
type Invoice struct {
	ID            int     `json:"id"`
	InvoiceNum    string  `json:"invoice_num,omitempty"`
	Date          string  `json:"date,omitempty"`
	DueDate       string  `json:"due_date,omitempty"`
	DatePaid      string  `json:"date_paid,omitempty"`
	Status        string  `json:"status,omitempty"`
	Subtotal      float64 `json:"subtotal,omitempty"`
	Credit        float64 `json:"credit,omitempty"`
	Tax           float64 `json:"tax,omitempty"`
	Total         float64 `json:"total,omitempty"`
	Currency      string  `json:"currency,omitempty"`
	PaymentMethod string  `json:"payment_method,omitempty"`
}

// invoicesListResponse unwraps {invoices, total}.
type invoicesListResponse struct {
	Invoices []Invoice `json:"invoices"`
	Total    int       `json:"total"`
}

// InvoicesList wraps GET /v1/invoices with optional status filter.
func (c *Client) InvoicesList(ctx context.Context, status string) ([]Invoice, error) {
	var q url.Values
	if status != "" {
		q = url.Values{"status": []string{status}}
	}
	var resp invoicesListResponse
	if err := c.Get(ctx, "/v1/invoices", q, &resp); err != nil {
		return nil, err
	}
	return resp.Invoices, nil
}

// InvoiceDetail is the response from GET /v1/invoices/{id}.
type InvoiceDetail struct {
	Invoice
	Items        []InvoiceItem        `json:"items,omitempty"`
	Transactions []InvoiceTransaction `json:"transactions,omitempty"`
}

type InvoiceItem struct {
	ID          int     `json:"id"`
	Description string  `json:"description"`
	Amount      float64 `json:"amount"`
	Type        string  `json:"type,omitempty"`
}

type InvoiceTransaction struct {
	ID            int     `json:"id"`
	Date          string  `json:"date,omitempty"`
	Gateway       string  `json:"gateway,omitempty"`
	Amount        float64 `json:"amount"`
	TransactionID string  `json:"transaction_id,omitempty"`
}

// InvoiceShow wraps GET /v1/invoices/{id}.
func (c *Client) InvoiceShow(ctx context.Context, id int) (*InvoiceDetail, error) {
	var inv InvoiceDetail
	if err := c.Get(ctx, "/v1/invoices/"+formatInt(id), nil, &inv); err != nil {
		return nil, err
	}
	return &inv, nil
}

// InvoicePayment is the response from POST /v1/invoices/{id}/pay.
//
// Amount is what actually left the balance, which is the invoice total
// unless credit had already been applied. It is a float64 to match every
// other money field in this package — see the note on Invoice; do not
// accumulate these values, compare and total them in minor units.
type InvoicePayment struct {
	InvoiceID int     `json:"invoice_id"`
	Amount    float64 `json:"amount"`
	Currency  string  `json:"currency,omitempty"`
	Message   string  `json:"message,omitempty"`
}

// InvoicePay wraps POST /v1/invoices/{id}/pay — pays an unpaid invoice
// from the account credit balance.
//
// This spends real money and the API offers no way to undo it. Callers
// that front a human (the CLI does) must confirm before calling.
//
// A 409 comes back as *Conflict, with two cases worth telling apart via
// APIError.Code: ALREADY_PAID (nothing to do) and INSUFFICIENT_BALANCE
// (top up first, see TopupCreate). A 404 comes back as *NotFound.
func (c *Client) InvoicePay(ctx context.Context, id int) (*InvoicePayment, error) {
	var out InvoicePayment
	if err := c.Post(ctx, "/v1/invoices/"+formatInt(id)+"/pay", nil, &out); err != nil {
		return nil, err
	}
	// The endpoint echoes invoice_id, but fall back to the id we asked
	// about so the result is never ambiguous about which invoice moved.
	if out.InvoiceID == 0 {
		out.InvoiceID = id
	}
	return &out, nil
}
